capsys breaks sys.stdin.encoding

Hi,

Using capsys breaks the use of sys.stdin.encoding.  (And probably does the same for sys.stdout.encoding and other file like object encoding attributes.)

Using sys.stdout.encoding is important when writing portable code that sometimes outputs stdout to the MS Windows console, which may use cp1252 encoding.  If you don't explicitly do something to encode your unicode strings and handle encoding failures then printing sometimes crashes your program with a UnicodeEncodeError exception.  But testing sys.stdout.encoding with capsys is awkward.

See the example.

```
$ venv/bin/pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
appdirs (1.4.3)
packaging (16.8)
pip (9.0.1)
py (1.4.33)
pyparsing (2.2.0)
pytest (3.0.7)
setuptools (35.0.1)
six (1.10.0)

Debian stable 8.7 (Jessie)
Python 2.7.9 (from Debian repo)
(You get the same output from Debian's Python 3.4.2.)
```

```python
#src/test_capsys.py 
from __future__ import print_function

import sys


def func(text):
    print(text.encode(sys.stdout.encoding, 'replace'))


def test_func(capsys):
    func('sample text')
    (out, err) = capsys.readouterr()
    assert out
    assert err == ''
```

```
$ venv/bin/pytest 
============================= test session starts ==============================
platform linux2 -- Python 2.7.9, pytest-3.0.7, py-1.4.33, pluggy-0.4.0
rootdir: *REDACTED*, inifile:
collected 1 items 

src/test_capsys.py F

=================================== FAILURES ===================================
__________________________________ test_func ___________________________________

capsys = <_pytest.capture.CaptureFixture instance at 0x7f90ba65a8c0>

    def test_func(capsys):
>       func('sample text')

src/test_capsys.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

text = 'sample text'

    def func(text):
>       print(text.encode(sys.stdout.encoding, 'replace'))
E       TypeError: encode() argument 1 must be string, not None

src/test_capsys.py:7: TypeError
=========================== 1 failed in 0.40 seconds ===========================
```

EDIT: fixed formatting

I could only reproduce this for py27 running it via: `tox -e py27,py34,py36,py35 -- testcases/test_capsys.py`. 
``` bash 
ERROR:   py27: commands failed
  py34: commands succeeded
  py36: commands succeeded
  py35: commands succeeded
```

The py27 outout is: 
``` bash
testcases/test_capsys.py F
==================================================================================================== short test summary info =====================================================================================================
FAIL testcases/test_capsys.py::test_func

============================================================================================================ FAILURES ============================================================================================================
___________________________________________________________________________________________________________ test_func ____________________________________________________________________________________________________________

capsys = <_pytest.capture.CaptureFixture instance at 0x7f4fc750d680>

    def test_func(capsys):
>       func('sample text')

testcases/test_capsys.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

text = 'sample text'

    def func(text):
>       print(text.encode(sys.stdout.encoding, 'replace'))
E       TypeError: encode() argument 1 must be string, not None

testcases/test_capsys.py:8: TypeError
==================================================================================================== 1 failed in 0.05 seconds ====================================================================================================
ERROR: InvocationError: '/home/maiksen/repos/pytest/.tox/py27/bin/pytest --lsof -rfsxX testcases/test_capsys.py'
```
 I am on the latest py3.4 available and on ubuntu 16.04.

--- a/AUTHORS
+++ b/AUTHORS
@@ -92,6 +92,7 @@ Kevin Cox
 Kodi B. Arfer
 Lee Kamentsky
 Lev Maximov
+Llandy Riveron Del Risco
 Loic Esteve
 Lukas Bednar
 Luke Murphy
@@ -165,3 +166,4 @@ Vitaly Lashmanov
 Vlad Dragos
 Wouter van Ackooy
 Xuecong Liao
+Zoltán Máté

--- a/_pytest/compat.py
+++ b/_pytest/compat.py
@@ -283,7 +283,16 @@ def _setup_collect_fakemodule():
 
 
 if _PY2:
-    from py.io import TextIO as CaptureIO
+    # Without this the test_dupfile_on_textio will fail, otherwise CaptureIO could directly inherit from StringIO.
+    from py.io import TextIO
+
+
+    class CaptureIO(TextIO):
+
+        @property
+        def encoding(self):
+            return getattr(self, '_encoding', 'UTF-8')
+
 else:
     import io
 

--- a/changelog/2375.trivial
+++ b/changelog/2375.trivial
@@ -1,0 +1,1 @@
+Provides encoding attribute on CaptureIO.

--- a/testing/test_capture.py
+++ b/testing/test_capture.py
@@ -1037,6 +1037,15 @@ def test_capture_not_started_but_reset():
     capsys.stop_capturing()
 
 
+def test_using_capsys_fixture_works_with_sys_stdout_encoding(capsys):
+    test_text = 'test text'
+
+    print(test_text.encode(sys.stdout.encoding, 'replace'))
+    (out, err) = capsys.readouterr()
+    assert out
+    assert err == ''
+
+
 @needsosdup
 @pytest.mark.parametrize('use', [True, False])
 def test_fdcapture_tmpfile_remains_the_same(tmpfile, use):

